### ISSUE TITLE
Running Deltaboost with Different Dataset error

### ISSUE THREAD

**Issue by diana-magalhaes** [Xtra-Computing/DeltaBoost#2]
### **Description**
I am experiencing a persistent `Segmentation fault (core dumped)` error when training on a custom dataset. The crash happens after the `generate_bin_trees_` function completes.
This occurs even when the custom dataset is clean, scaled, and uses a configuration file with parameters identical to the working covtype example. The issue has been reproduced with both a multi-class and a binary version of my dataset.
### **Log Output**
The program always crashes after the following log messages:
**Log**
```
2025-08-26 10:46:36,951 INFO deltaboost.cpp:38 : starting building trees
2025-08-26 10:46:37,612 INFO hist_cut.cpp:706 : generate_bin_trees_ time: 0s
Segmentation fault (core dumped)
```
### Debugging Steps Performed
1. Identical Configuration: The program still crashes when using a configuration file where all hyperparameters (max_bin_size, min_diff_gain, etc.) are identical to the ones in the working covtype example.
2. Data Scaling: The custom dataset has been Min-Max scaled, with all feature values in the [0, 1] range.
3. Binary vs. Multi-class: The crash occurs with both a multi-class version of the dataset (using objective=multi:softmax) and a binary version (using objective=binary:logistic).
4. Data Integrity Check 1 (Constant Columns): I have analyzed the dataset and confirmed that there are no feature columns where all values are zero or otherwise constant.
5. Data Integrity Check 2 (Single Unique Value): I have also confirmed that there are no feature columns that contain only a single unique non-zero value.

**Comment by JerryLife** [comment]
Thank you for your interest in DeltaBoost. Though not a theoretical challenge, the current implementation of DeltaBoost does not support multi-class classification.

Could you provide a link to a tiny binary-classification dataset that cause this issue for our reproduction? To help identify the bug, it is also recommended to test whether this dataset works for the standalone version of [FedTree](https://github.com/Xtra-Computing/FedTree) which DeltaBoost is built on.

**Comment by diana-magalhaes** [comment]
I’ve created a repository with the dataset I’m using. The dataset is already preprocessed and ready for use: [GeNIS](https://github.com/diana-magalhaes/genis-dataset). I’m currently checking if it can be used with FedTree.

**Comment by diana-magalhaes** [comment]
I just tested my dataset on FedTree, and it does not work. It gives a segmentation fault error just like with DeltaBoost.

**Comment by JerryLife** [comment]
Thank you for your feedback. Our team is looking into this issue. However, since this issue stems from a bug of base repository, it may require a significant update and take longer time to fix. Thank you for your understanding and patience.

**Comment by JerryLife** [comment]
When reproducing the issue, I find the provided `GeNIS` dataset is in CSV. The required format of DeltaBoost obtained from `download_datasets.sh` is [LIBSVM](https://stats.stackexchange.com/questions/61328/libsvm-data-format) format, which looks like below:
```
-1 1:-500 2:128 3:0.140625 4:0.304688 5:0.234375 6:0.129771 7:0.328244 8:0.229008
-1 1:-541 2:123 3:0.140625 4:0.304688 5:0.234375 6:0.146341 7:0.333333 8:0.211382
-1 1:-479 2:124 3:0.140625 4:0.304688 5:0.234375 6:0.137097 7:0.322581 8:0.233871
-1 1:-427 2:128 3:0.140625 4:0.304688 5:0.234375 6:0.153846 7:0.392308 8:0.184615
```
May I confirm if you have converted the format to LIBSVM before running？If so, could you please also provide the specific process of running `genis` dataset?

**Comment by diana-magalhaes** [comment]
Hello, yes, I did transform the CSV files into LIBSVM format.  I submitted to the repository I mentioned previously the code I used to transform the CSV files into LIBSVM, as well as the files themselves, which I used to run DeltaBoost.

**Comment by JerryLife** [comment]
Thank you for your information and it greatly expedite the fix.

Please refer to the [latest commit](https://github.com/Xtra-Computing/DeltaBoost/commit/91524201a5694600ab1d83d667c72fa67cc12dd5) for the updated code which should **fix the segfault issue for genis dataset**. The reason is a combination of format issue, imbalanced classes that lead to single label in the deleting set, feature with single value, etc. After all, DeltaBoost now works on genis dataset with either csv or libsvm format.

Please kindly be aware that there is a known issue that one may still occasionally encounter a Segfault in tree initialization, probably due to paralism or randomness. A simple workaround is to rerun the script and it should work properly within 3 retries. We are still working on it but it cannot be easily fixed. As this occasional segfault is not a dataset-specific issue, I will close this issue for resolved.
